File: Services/Concrete/Excel/Sheets/Multiple/KillsSheet.cs
using System.Collections.Generic;
using Core.Models;

namespace Services.Concrete.Excel.Sheets.Multiple
{
    internal class KillsSheet: MultipleDemoSheet
    {
        private readonly Dictionary<string, KillSheetRow[]> _rowsPerDemoId = new Dictionary<string, KillSheetRow[]>();

        protected override string GetName()
        {
            return "Kills";
        }

        protected override string[] GetColumnNames()
        {
            return new[]
            {

                "Demo ID",
                "Tick",
                "Round",
                "Time death (s)",
                "Killer",
                "Killer SteamID",
                "Killer side",
                "Killer team",
                "Killer bot",
                "Killer blinded",
                "Killer vel X",
                "Killer vel Y",
                "Killer vel Z",
                "Victim",
                "Victim SteamId",
                "Victim side",
                "Victim team",
                "Victim bot",
                "Victim blinded",
                "Assister",
                "Assister SteamID",
                "assister bot",
                "Weapon",
                "Headshot",
                "Crouching",
                "Trade kill",
                "Killer X",
                "Killer Y",
                "Killer Z",
                "Victim X",
                "Victim Y",
                "Victim Z",
            };
        }

        public KillsSheet(Workbook workbook): base(workbook)
        {
        }

        public override void AddDemo(Demo demo)
        {
            if (!_rowsPerDemoId.ContainsKey(demo.Id))
            {
                _rowsPerDemoId[demo.Id] = new KillSheetRow[demo.Kills.Count];
                for(var index = 0; index < demo.Kills.Count; index++)
                {
                    var kill = demo.Kills[index];
                    var row = new KillSheetRow
                    {
                        Tick = kill.Tick,
                        RoundNumber = kill.RoundNumber,
                        TimeDeathSeconds = kill.TimeDeathSeconds,
                        KillerName = kill.KillerName,
                        KillerSteamId = kill.KillerSteamId,
                        KillerSide = kill.KillerSide,
                        KillerTeamName = kill.KillerTeam,
                        KillerIsControllingBot = kill.IsKillerBot,
                        KillerIsBlinded = kill.KillerIsBlinded,
                        KillerVelocityX = kill.KillerVelocityX,
                        KillerVelocityY = kill.KillerVelocityY,
                        KillerVelocityZ = kill.KillerVelocityZ,
                        VictimName = kill.KilledName,
                        VictimSteamId = kill.KilledSteamId,
                        VictimSide = kill.KilledSide,
                        VictimTeamName = kill.KilledTeam,
                        VictimIsControllingBot = kill.KilledIsControllingBot,
                        VictimIsBlinded = kill.VictimIsBlinded,
                        AssisterName = kill.AssisterName,
                        AssisterSteamId = kill.AssisterSteamId,
                        AssisterIsControllingBot = kill.AssisterIsControllingBot,
                        WeaponName = kill.Weapon.Name,
                        IsHeadshot = kill.IsHeadshot,
                        IsKillerCrouching = kill.IsKillerCrouching,
                        IsTradeKill = kill.IsTradeKill,
                        KillerX = kill.Point.KillerX,
                        KillerY = kill.Point.KillerY,
                        KillerZ = kill.Point.KillerZ,
                        VictimX = kill.Point.VictimX,
                        VictimY = kill.Point.VictimY,
                        VictimZ = kill.Point.VictimZ,
                    };

                    _rowsPerDemoId[demo.Id][index] = row;
                }
            }
        }

        public override void Generate()
        {
            foreach (var entry in _rowsPerDemoId)
            {
                foreach (var kill in entry.Value)
                {
                    var cells = new List<object>
                    {
                        entry.Key,
                        kill.Tick,
                        kill.RoundNumber,
                        kill.TimeDeathSeconds,
                        kill.KillerName,
                        kill.KillerSteamId.ToString(),
                        kill.KillerSide.AsString(),
                        kill.KillerTeamName,
                        kill.KillerIsControllingBot,
                        kill.KillerIsBlinded,
                        kill.KillerVelocityX,
                        kill.KillerVelocityY,
                        kill.KillerVelocityZ,
                        kill.VictimName,
                        kill.VictimSteamId.ToString(),
                        kill.VictimSide.AsString(),
                        kill.VictimTeamName,
                        kill.VictimIsControllingBot,
                        kill.VictimIsBlinded,
                        kill.AssisterName,
                        kill.AssisterSteamId.ToString(),
                        kill.AssisterIsControllingBot,
                        kill.WeaponName,
                        kill.IsHeadshot,
                        kill.IsKillerCrouching,
                        kill.IsTradeKill,
                        kill.KillerX,
                        kill.KillerY,
                        kill.KillerZ,
                        kill.VictimX,
                        kill.VictimY,
                        kill.VictimZ,
                    };
                    WriteRow(cells);
                }
            }
        }
    }
}

File: src/node/xlsx/match-export/kills-sheet.ts
import type { Kill } from 'csdm/common/types/kill';
import type { Column } from '../column';
import { SingleMatchExportSheet } from './single-match-export-sheet';

export class KillsSheet extends SingleMatchExportSheet<Kill> {
  protected getName() {
    return 'Kills';
  }

  protected getColumns(): Column<Kill>[] {
    return [
      {
        name: 'round_number',
        cellFormatter: (row) => row.roundNumber,
      },
      {
        name: 'tick',
        cellFormatter: (row) => row.tick,
      },
      {
        name: 'frame',
        cellFormatter: (row) => row.frame,
      },
      {
        name: 'killer_steam_id',
        cellFormatter: (row) => row.killerSteamId,
      },
      {
        name: 'killer_name',
        cellFormatter: (row) => row.killerName,
      },
      {
        name: 'killer_side',
        cellFormatter: (row) => row.killerSide,
      },
      {
        name: 'victim_steam_id',
        cellFormatter: (row) => row.victimSteamId,
      },
      {
        name: 'victim_name',
        cellFormatter: (row) => row.victimName,
      },
      {
        name: 'victim_side',
        cellFormatter: (row) => row.victimSide,
      },
      {
        name: 'assister_steam_id',
        cellFormatter: (row) => row.assisterSteamId,
      },
      {
        name: 'assister_name',
        cellFormatter: (row) => row.assisterName,
      },
      {
        name: 'assister_side',
        cellFormatter: (row) => row.assisterSide,
      },
      {
        name: 'weapon_name',
        cellFormatter: (row) => row.weaponName,
      },
      {
        name: 'penetrated_objects',
        cellFormatter: (row) => row.penetratedObjects,
      },
      {
        name: 'distance',
        cellFormatter: (row) => row.distance,
      },
      {
        name: 'is_headshot',
        cellFormatter: (row) => row.isHeadshot,
      },
      {
        name: 'is_assisted_flash',
        cellFormatter: (row) => row.isAssistedFlash,
      },
      {
        name: 'is_trade_kill',
        cellFormatter: (row) => row.isTradeKill,
      },
      {
        name: 'is_trade_death',
        cellFormatter: (row) => row.isTradeDeath,
      },
      {
        name: 'is_no_scope',
        cellFormatter: (row) => row.isNoScope,
      },
      {
        name: 'is_through_smoke',
        cellFormatter: (row) => row.isThroughSmoke,
      },
      {
        name: 'is_killer_airbone',
        cellFormatter: (row) => row.isKillerAirborne,
      },
      {
        name: 'is_victim_airbone',
        cellFormatter: (row) => row.isVictimAirborne,
      },
      {
        name: 'is_killer_blinded',
        cellFormatter: (row) => row.isKillerBlinded,
      },
      {
        name: 'is_victim_blinded',
        cellFormatter: (row) => row.isVictimBlinded,
      },
      {
        name: 'is_killer_controlling_bot',
        cellFormatter: (row) => row.isKillerControllingBot,
      },
      {
        name: 'is_victim_controlling_bot',
        cellFormatter: (row) => row.isVictimControllingBot,
      },
      {
        name: 'is_assister_controlling_bot',
        cellFormatter: (row) => row.isAssisterControllingBot,
      },
      {
        name: 'killer_x',
        cellFormatter: (row) => row.killerX,
      },
      {
        name: 'killer_y',
        cellFormatter: (row) => row.killerY,
      },
      {
        name: 'killer_z',
        cellFormatter: (row) => row.killerZ,
      },
      {
        name: 'victim_x',
        cellFormatter: (row) => row.victimX,
      },
      {
        name: 'victim_y',
        cellFormatter: (row) => row.victimY,
      },
      {
        name: 'victim_z',
        cellFormatter: (row) => row.victimZ,
      },
      {
        name: 'assister_x',
        cellFormatter: (row) => row.assisterX,
      },
      {
        name: 'assister_y',
        cellFormatter: (row) => row.assisterY,
      },
      {
        name: 'assister_z',
        cellFormatter: (row) => row.assisterZ,
      },
    ];
  }

  public generate() {
    for (const kill of this.match.kills) {
      this.writeRow(kill);
    }
  }
}
